### PR TITLE
HHH-17972 - Allow changing tenant identifier in SharedSessionBuilder

### PR DESCRIPTION
**Pull request by jrenaat**
https://hibernate.atlassian.net/browse/HHH-17972

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 1]
```diff
@@ -160,7 +160,7 @@ public abstract class AbstractSharedSessionContract implements SharedSessionCont
 
 	private final Interceptor interceptor;
 
-	private final Object tenantIdentifier;
+	protected Object tenantIdentifier;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:163]: Please revert this.

File: hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java [hunk 2]
```diff
@@ -2063,14 +2063,20 @@ private SharedSessionBuilderImpl(SessionImpl session) {
 
 		@Override
 		public SharedSessionBuilderImpl tenantIdentifier(String tenantIdentifier) {
-			// todo : is this always true?  Or just in the case of sharing JDBC resources?
-			throw new SessionException( "Cannot redefine tenant identifier on child session" );
+			if ( !session.getSessionFactory().getSessionFactoryOptions().isMultiTenancyEnabled() ) {
+				throw new SessionException( "Cannot redefine tenant identifier on child session" );
+			}
+			session.tenantIdentifier = session.getSessionFactory().getTenantIdentifierJavaType().toString( tenantIdentifier );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java:2069]:
```suggestion
			this.tenantIdentifier = session.getSessionFactory().getTenantIdentifierJavaType().toString( tenantIdentifier );
```

File: hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java [hunk 3]
```diff
@@ -2063,14 +2063,20 @@ private SharedSessionBuilderImpl(SessionImpl session) {
...
 		}
 
 		@Override
 		public SharedSessionBuilderImpl tenantIdentifier(Object tenantIdentifier) {
-			// todo : is this always true?  Or just in the case of sharing JDBC resources?
-			throw new SessionException( "Cannot redefine tenant identifier on child session" );
+			if ( session.getSessionFactory().getSessionFactoryOptions().isMultiTenancyEnabled() ) {
+				throw new SessionException( "Cannot redefine tenant identifier on child session" );
+			}
+			session.tenantIdentifier = tenantIdentifier;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java:2078]:
```suggestion
			this.tenantIdentifier = tenantIdentifier;
```

File: hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java (hunk 2 above)
Comment (jrenaat) [reply hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java:2069]: Doing it lika so has the side-effect that now the insert after setting the new tenant id now also doesn't reflect the newly set id, it's just creates another instance with the old tenant id

File: hibernate-core/src/main/java/org/hibernate/SharedSessionBuilder.java [hunk 4]
```diff
@@ -67,6 +67,8 @@ public interface SharedSessionBuilder extends SessionBuilder {
 	 */
 	SharedSessionBuilder autoClose();
 
+	SharedSessionBuilder setTenantIdentifier(Object tenantIdentifier);
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/SharedSessionBuilder.java:70]: @jrenaat after 8ceed5ab5a0d434086e8b3760fdf3fa9e5d168c4 we should not need this new method in `SharedSessionBuilder`.

**Review by beikov — APPROVED** [review]

**Comment by gavinking** [comment]
yay!

**Pull request merged**